[LBSE] Implement support for patterns

rwlbuis opened the pull request:
#### db4c995fc7b47368c3f67671b89c42dc632402a7
<pre>
[LBSE] Implement support for patterns
<a href="https://bugs.webkit.org/show_bug.cgi?id=267041">https://bugs.webkit.org/show_bug.cgi?id=267041</a>

Reviewed by Nikolas Zimmermann.

Implement pattern support using a different approach to legacy including the
use of RenderLayer::paintSVGLayer ro paint the pattern contents. Right now
there is no caching of pattern tiles (unlike in legacy) at all to keep things simple.

Note that there are still various pattern test failures due to problems with hidpi.

Adding convenient pattern API for stroke painting in GraphicsContext is left for later (see SVGInlineTextBox.cpp).

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-03-b-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/batik/paints/patternRegionA-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/absolute-root-position-masking-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/absolute-sized-content-with-resources-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-update-pattern-child-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-update-pattern-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/oversized-pattern-scale-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-cycle-detection-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-cycle-detection-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-excessive-malloc-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-in-defs-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-no-pixelation-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-no-pixelation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-rotate-gaps-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-size-bigger-than-target-size-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-y-offset-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-y-offset-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pending-resource-after-removal-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/recursive-pattern-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/pattern-object-bounding-box-update-pattern-content-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/pattern-user-space-on-use-update-pattern-content-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/pattern-no-pixelation-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/pattern-skew-transformed-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/pattern-y-offset-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/recursive-pattern-expected.png:
* LayoutTests/svg/repaint/pattern-object-bounding-box-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box-transformed.html: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box-update-pattern-content-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box-update-pattern-content.html: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box.html: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-transformed.html: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-update-pattern-content-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-update-pattern-content.html: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::ReferencedSVGResources::referencedSVGResourceIDs):
(WebCore::ReferencedSVGResources::referencedPaintServerElement):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::svgFillPaintServerResourceFromStyle const):
(WebCore::RenderLayerModelObject::svgStrokePaintServerResourceFromStyle const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderSVGResourcePaintServer const):
(WebCore::RenderObject::isRenderSVGResourcePattern const):
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp: Added.
(WebCore::RenderSVGResourcePattern::RenderSVGResourcePattern):
(WebCore::RenderSVGResourcePattern::collectPatternAttributesIfNeeded):
(WebCore::clear2DRotation):
(WebCore::RenderSVGResourcePattern::buildPattern):
(WebCore::RenderSVGResourcePattern::prepareFillOperation):
(WebCore::RenderSVGResourcePattern::prepareStrokeOperation):
(WebCore::RenderSVGResourcePattern::buildTileImageTransform const):
(WebCore::RenderSVGResourcePattern::createTileImage const):
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.h: Added.
(WebCore::RenderSVGResourcePattern::invalidatePattern):
* Source/WebCore/rendering/svg/RenderSVGResourcePatternInlines.h: Added.
(WebCore::RenderSVGResourcePattern::patternElement const):
(WebCore::calculatePatternBoundaries):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::acquirePaintingResource):
(WebCore::SVGInlineTextBox::releasePaintingResource):
(WebCore::SVGInlineTextBox::paintTextWithShadows):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGHitTestCycleDetectionScope::SVGHitTestCycleDetectionScope):
(WebCore::SVGHitTestCycleDetectionScope::~SVGHitTestCycleDetectionScope):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourcePattern.cpp:
(WebCore::LegacyRenderSVGResourcePattern::LegacyRenderSVGResourcePattern):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):
* Source/WebCore/svg/SVGPatternElement.cpp:
(WebCore::SVGPatternElement::svgAttributeChanged):
(WebCore::SVGPatternElement::createElementRenderer):

Canonical link: <a href="https://commits.webkit.org/273757@main">https://commits.webkit.org/273757@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f20df524ed2bd9d583dcd4f2abb5579952499848

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36568 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38790 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39231 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32793 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37802 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17981 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12588 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31408 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37129 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13072 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32347 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11443 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11461 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32606 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40476 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33137 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32946 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37391 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11709 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9550 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35511 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13394 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/32265 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8290 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12134 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12593 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -24326,6 +24336,14 @@
 			path = ios;
 			sourceTree = "<group>";
 		};
+		44E72D2D2B445D77009CA525 /* Recovered References */ = {
```
What is this about? maybe it is trying to correct a mistake I made earlier in the gradient PR?

rwlbuis replied:
Now removed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,225 @@
+    while (current) {
+        current->collectPatternAttributes(attributes);
+
+        auto target = SVGURIReference::targetElementFromIRIString(current->href(), current->treeScopeForSVGReferences());
```
This is not necessary for gradients (compare to collectGradientAttributesIfNeeded) - because the underlying SVGLinearGradientElement::collectGradientAttributes() method handles it, and SVGPatternElement::collectPatternAttributes() not.

Your code would currently lead to endless cycles on crafted SVG content - nobody breaks them.
I would suggest to move responsibility into DOM for patterns as well, especially to support the special chaining mode via (xlink:)href.

nikolaszimmermann replied:
This is still a concern and source of confusion- we should move more logic into collectPatternAttrs as for gradients...

nikolaszimmermann replied:
Not so confident anymore, maybe we should model gradients more like patterns? This needs discussion, but the cleanup can be done later.

rwlbuis replied:
I think patterns have a different kind of recursion which is hard to combine with collectFoo, i.e. the pattern content can reference other patterns, this can't happen with gradients. But yes, can be visited in the future.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,225 @@
+    if (!buildTileImageTransform(renderer, *m_attributes, patternElement(), tileBoundaries, tileImageTransform))
+        return nullptr;
+
+    auto absoluteTransform = SVGRenderingContext::calculateTransformationToOutermostCoordinateSystem(renderer);
```
Beware, this is legacy code. In LBSE this won't give correct answers. You probably need to use SVGLayerTransformComputation directly:

```
auto transform = SVGLayerTransformComputation(child).computeAccumulatedTransform(&renderer, TransformState::TrackSVGCTMMatrix);
```

since you do need access to the CTM, to extract the xScale/yScale.

The `float calculateScreenFontSizeScalingFactor() const` helper function in SVGLayerTransformComputation has the same purpose (it just returns `sqrt((xScale^2+yScale^2)/2`). You could add a helper to return a FloatSize with xScale, yScale, or a std::tuple<float, float>, so one could use structured-binding and use something like "auto [xScale, yScale] = SVGLayerTransformComputation(..).calculateXYCTMScaleForRenderer(..)".

Whatever you prefer :-)

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,225 @@
+
+bool RenderSVGResourcePattern::prepareFillOperation(GraphicsContext& context, const RenderLayerModelObject& targetRenderer, const RenderStyle& style)
+{
+    if (SVGHitTestCycleDetectionScope::isVisiting(*this))
```
That is not necessary for gradients, as cycles are solved differently, no? I think once you fix collectPatternAttributes(), this is no longer necessary. Should be checked.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,225 @@
+    return true;
+}
+
+static inline FloatRect calculatePatternBoundaries(const PatternAttributes& attributes, const FloatRect& objectBoundingBox, const SVGPatternElement& patternElement)
```
Should be moved into RenderSVGResourcePatternInlines.h

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,225 @@
+
+    auto tileSize = roundedUnscaledImageBufferSize(size, scale);
+
+    // FIXME: Use createImageBuffer(rect, scale), delete the above calculations and fix 'tileImageTransform'
```
Is that yours?

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,225 @@
+    for (auto& child : childrenOfType<SVGElement>(*attributes.patternContentElement())) {
+        if (!child.renderer())
+            continue;
+        SVGRenderingContext::renderSubtreeToContext(tileImageContext, *child.renderer(), contentTransformation);
```
That is all legacy stuff, SVGRenderingContext does not work for LBSE. Transformations would be broken, rendering at wrong spots -- you need to paint/hit-test via layers, not using the render tree.

Excerpt from downstream RenderSVGResourcePattern:

```
    auto* patternRenderer = static_cast<RenderSVGResourcePattern*>(attributes.patternContentElement()->renderer());
    ASSERT(patternRenderer);
    ASSERT(patternRenderer->hasLayer());
    
    // Draw the content into the ImageBuffer.
    patternRenderer->layer()->paintSVGResourceLayer(tileImageContext, LayoutRect::infiniteRect(), tileImageTransform);
```

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
```diff
@@ -669,6 +669,17 @@ void SVGInlineTextBox::paintTextWithShadows(GraphicsContext& context, const Rend
+                    usedContext->setCompositeOperation(CompositeOperator::DestinationIn);
+                    usedContext->beginTransparencyLayer(1);
+                } else if (usedContext->strokePattern()) {
+                    usedContext->setFillPattern(*usedContext->strokePattern());
```
I think you also want to extend GraphicsContext :-) a `fillRect(const FloatRect&, Pattern&)` variant, would allow you to avoid the state mutation here. (Note: you also need to assure it's reset correctly, the current restore() time is too lat probably)

nikolaszimmermann reviewed (commented):
Great start! Needs to move away from some pre-LBSE idioms that involve SVGRenderingContext.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2839,6 +2839,8 @@ void RenderLayer::paintSVGResourceLayer(GraphicsContext& context, GraphicsContex
     auto localPaintDirtyRect = LayoutRect::infiniteRect();
 
     auto* rootPaintingLayer = enclosingSVGRootLayer();
+    if (is<RenderSVGHiddenContainer>(parent()->renderer()))
```
Can you explain that?

rwlbuis replied:
Yeah this is tricky. It fixes custom/pattern-userSpaceOnUse-userToBaseTransform.xhtml.
The outer svg's have a 1px border which is mistakenly taken into account (before that change) and offsets the pattern contents by 1x1, showing the red rectangle.

nikolaszimmermann replied:
I think it deserves a comment.

rwlbuis replied:
Actually I think I'll leave out this part, it breaks recursive-mask.svg and maybe some others. We'll have to visit this 1px border problem later...

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3490,7 +3492,8 @@ void RenderLayer::paintLayerByApplyingTransform(GraphicsContext& context, const
     LayerPaintingInfo transformedPaintingInfo(paintingInfo);
     transformedPaintingInfo.rootLayer = this;
-    transformedPaintingInfo.paintDirtyRect = LayoutRect(encloseRectToDevicePixels(valueOrDefault(transform.inverse()).mapRect(paintingInfo.paintDirtyRect), deviceScaleFactor));
+    if (!transformedPaintingInfo.paintDirtyRect.isInfinite())
```
As we use InfiniteRect as input when painting pattern content right?

rwlbuis replied:
Right, maybe a std::optional would have been better.  Anyway this was already a fix in the mask/clip PR from yesterday which landed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,226 @@
+    patternSpaceTransform.translate(tileBoundaries.location());
+    patternSpaceTransform.scale(tileBoundaries.size() / tileImageSize);
+
+    AffineTransform patternTransform = m_attributes->patternTransform();
```
Auto

rwlbuis replied:
Fixed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,226 @@
+
+bool RenderSVGResourcePattern::buildPatternIfNeeded(GraphicsContext& context, const RenderLayerModelObject& targetRenderer)
+{
+    collectPatternAttributesIfNeeded();
```
I think for gradients this returns a bool, false indicating early exit due to empty bbox- why not mirror it 1:1 here

rwlbuis replied:
I made collectPatternAttributesIfNeeded do more checks.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,226 @@
+
+bool RenderSVGResourcePattern::prepareFillOperation(GraphicsContext& context, const RenderLayerModelObject& targetRenderer, const RenderStyle& style)
+{
+    if (SVGHitTestCycleDetectionScope::isVisiting(*this))
```
I think this is not needed if you resolve cycles during collectPatternAttr - or am I mistaken?

rwlbuis replied:
I believe putting all logic in collectPatternAttr on DOM side will not catch some cases from custom/pattern-content-inheritance-cycle.svg like:
<!-- a => b => a -->
        <pattern id="a" x="0" y="0" width=".25" height=".25">
            <rect fill="lime" width="100%" height="100%"/>
            <rect fill="url(#b)" width="100%" height="100%"/>
        </pattern>
        <pattern id="b" xlink:href="#a"/>
At least when I tried it crashed :)

nikolaszimmermann replied:
Ok fair enough

nikolaszimmermann replied:
I just wanted to suggest to unify the code with gradient, to avoid the URI resolving in collectGradientAttributes, but...

I think, only patterns have the ability to have children that can reference other resources - gradients only have stop children.

nikolaszimmermann replied:
That would mean everything is fine, we only need to check the coverage for cycles and gradients is sufficient.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,226 @@
+static AffineTransform transformOnNonScalingStroke(const RenderObject& object, const AffineTransform& resourceTransform)
+{
+    SVGGraphicsElement* element = downcast<SVGGraphicsElement>(object.node());
+    AffineTransform transform = element->getScreenCTM(SVGLocatable::DisallowStyleUpdate);
```
Is that how RenderSVGShape computes the nonScalingStroke trafo? Should be unified

nikolaszimmermann replied:
```
AffineTransform RenderSVGShape::nonScalingStrokeTransform() const
{       
    return graphicsElement().getScreenCTM(SVGLocatable::DisallowStyleUpdate);
}   
```

The incoming RenderObject is a RenderSVGShape, so you can just invoke nonScalingStrokeTransform() instead of duplicating the logic.

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,226 @@
+
+bool RenderSVGResourcePattern::buildTileImageTransform(const RenderElement& renderer, const PatternAttributes& attributes, const SVGPatternElement& patternElement, FloatRect& patternBoundaries, AffineTransform& tileImageTransform) const
+{
+    FloatRect objectBoundingBox = renderer.objectBoundingBox();
```
You can fold that in the viewBox case where this variable is needed..

rwlbuis replied:
But objectBoundingBox is used in the line right after it is defined.

nikolaszimmermann replied:
Oh missed that. Then only some autos could help here.

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
```diff
@@ -659,13 +660,23 @@ void SVGInlineTextBox::paintTextWithShadows(GraphicsContext& context, const Rend
+                    if (gradient)
+                        usedContext->fillRect(textRootBlock->repaintRectInLocalCoordinates(), *gradient, gradientSpaceTransform);
+                    else {
+                        GraphicsContextStateSaver stateSaver(*usedContext);
```
I would then check if a fillRect(FloatRect,Pattern&) exists to avoid that state swap

nikolaszimmermann replied:
Do you plan to take this in a separated PR?

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGRenderingContext.h:
```diff
@@ -74,6 +74,8 @@ class SVGRenderingContext {
     // Support for the buffered-rendering hint.
     bool bufferForeground(RefPtr<ImageBuffer>&);
 
+    static void clear2DRotation(AffineTransform&);
```
Please no new stuff in that file - can you keep it in the pattern impl?

rwlbuis replied:
Sure, looks like the pattern impl may be the only user of clear2DRotation anyway, fixed.

nikolaszimmermann reviewed (commented):
Great patch, almost r+. New repaint tests would be great too

nikolaszimmermann reviewed (commented):
I think only fillRect() +  Pattern variant needs discussion the rest looks good.

nikolaszimmermann commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -21339,6 +21345,7 @@
 				0867D69AFE84028FC02AAC07 /* Frameworks */,
 				034768DFFF38A50411DB9C8B /* Products */,
 				1CDD44660BA9C80000F90147 /* Configurations */,
+				44E72D2D2B445D77009CA525 /* Recovered References */,
```
I think this can be omitted.

rwlbuis replied:
Ah, yeah some leftover, removed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
```
Technically also 2021, 2022, and 2023.

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,231 @@
+
+    auto patternTransform = m_attributes->patternTransform();
+    if (!patternTransform.isIdentity())
+        patternSpaceTransform = patternTransform * patternSpaceTransform;
```
we usually call multiply instead of invoking operators, no?

rwlbuis replied:
Yes, I have not used these APIs in a long time :-) Fixed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,231 @@
+    context.setAlpha(svgStyle.strokeOpacity());
+    SVGRenderSupport::applyStrokeStyleToContext(context, style, targetRenderer);
+    if (svgStyle.vectorEffect() == VectorEffect::NonScalingStroke) {
+        if (is<RenderSVGShape>(targetRenderer))
```
The two if conditions should be unified into one to make this slightly better readable.

rwlbuis replied:
I think that does not really work:
if (auto* shape = dynamicDowncast<RenderSVGShape>(targetRenderer) ; svgStyle.vectorEffect() == VectorEffect::NonScalingStroke) {

Then we may set shape without using it.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,231 @@
+RefPtr<ImageBuffer> RenderSVGResourcePattern::createTileImage(const PatternAttributes& attributes, const FloatRect& tileBoundaries, const FloatRect& absoluteTileBoundaries, const AffineTransform& tileImageTransform) const
+{
+    auto clampedAbsoluteTileBoundaries = ImageBuffer::clampedRect(absoluteTileBoundaries);
+    auto tileImage = ImageBuffer::create(roundedIntSize(clampedAbsoluteTileBoundaries.size()), RenderingPurpose::Unspecified, 1, DestinationColorSpace::SRGB(), PixelFormat::BGRA8);
```
At some point we need to revisit the whole color space handling, not sure how 'color-interpolation' influences e.g. patterns.

rwlbuis replied:
I left a FIXME.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
```diff
@@ -0,0 +1,231 @@
+    ASSERT(patternRenderer);
+    ASSERT(patternRenderer->hasLayer());
+
+    // We only want to detect cycles for patternRenderer != this, since paintSVGResourceLayer itself
```
When does that happen? referenced patterns via href?

rwlbuis replied:
It seemed to happen in Debug with a simple case, like svg/repaint/pattern-object-bounding-box-transformed.html.

nikolaszimmermann reviewed (approved):
Great, r=me with a few comments left:

nikolaszimmermann reviewed (approved):
r=me

rwlbuis committed:
[LBSE] Implement support for patterns
https://bugs.webkit.org/show_bug.cgi?id=267041

Reviewed by Nikolas Zimmermann.

Implement pattern support using a different approach to legacy including the
use of RenderLayer::paintSVGLayer ro paint the pattern contents. Right now
there is no caching of pattern tiles (unlike in legacy) at all to keep things simple.

Note that there are still various pattern test failures due to problems with hidpi.

Adding convenient pattern API for stroke painting in GraphicsContext is left for later (see SVGInlineTextBox.cpp).

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-03-b-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/batik/paints/patternRegionA-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/absolute-root-position-masking-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/absolute-sized-content-with-resources-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-update-pattern-child-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-update-pattern-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/oversized-pattern-scale-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-cycle-detection-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-cycle-detection-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-excessive-malloc-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-in-defs-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-no-pixelation-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-no-pixelation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-rotate-gaps-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-size-bigger-than-target-size-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-y-offset-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-y-offset-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pending-resource-after-removal-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/recursive-pattern-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/pattern-object-bounding-box-update-pattern-content-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/pattern-user-space-on-use-update-pattern-content-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/pattern-no-pixelation-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/pattern-skew-transformed-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/pattern-y-offset-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/recursive-pattern-expected.png:
* LayoutTests/svg/repaint/pattern-object-bounding-box-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box-transformed.html: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box-update-pattern-content-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box-update-pattern-content.html: Added.
* LayoutTests/svg/repaint/pattern-object-bounding-box.html: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-transformed.html: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-update-pattern-content-expected.txt: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use-update-pattern-content.html: Added.
* LayoutTests/svg/repaint/pattern-user-space-on-use.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::ReferencedSVGResources::referencedSVGResourceIDs):
(WebCore::ReferencedSVGResources::referencedPaintServerElement):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::svgFillPaintServerResourceFromStyle const):
(WebCore::RenderLayerModelObject::svgStrokePaintServerResourceFromStyle const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderSVGResourcePaintServer const):
(WebCore::RenderObject::isRenderSVGResourcePattern const):
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp: Added.
(WebCore::RenderSVGResourcePattern::RenderSVGResourcePattern):
(WebCore::RenderSVGResourcePattern::collectPatternAttributesIfNeeded):
(WebCore::clear2DRotation):
(WebCore::RenderSVGResourcePattern::buildPattern):
(WebCore::RenderSVGResourcePattern::prepareFillOperation):
(WebCore::RenderSVGResourcePattern::prepareStrokeOperation):
(WebCore::RenderSVGResourcePattern::buildTileImageTransform const):
(WebCore::RenderSVGResourcePattern::createTileImage const):
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.h: Added.
(WebCore::RenderSVGResourcePattern::invalidatePattern):
* Source/WebCore/rendering/svg/RenderSVGResourcePatternInlines.h: Added.
(WebCore::RenderSVGResourcePattern::patternElement const):
(WebCore::calculatePatternBoundaries):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::acquirePaintingResource):
(WebCore::SVGInlineTextBox::releasePaintingResource):
(WebCore::SVGInlineTextBox::paintTextWithShadows):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGHitTestCycleDetectionScope::SVGHitTestCycleDetectionScope):
(WebCore::SVGHitTestCycleDetectionScope::~SVGHitTestCycleDetectionScope):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourcePattern.cpp:
(WebCore::LegacyRenderSVGResourcePattern::LegacyRenderSVGResourcePattern):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):
* Source/WebCore/svg/SVGPatternElement.cpp:
(WebCore::SVGPatternElement::svgAttributeChanged):
(WebCore::SVGPatternElement::createElementRenderer):

Canonical link: https://commits.webkit.org/273757@main
